mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/RowMatrix.scala

```diff
@@ -439,7 +439,8 @@ class RowMatrix @Since("1.0.0") (
       "  Cannot compute the covariance of a RowMatrix with <= 1 row.")
     val mean = Vectors.fromML(summary.mean)
 
-    if (rows.first().isInstanceOf[DenseVector]) {
+    // If all the rows are sparse vectors, then compute based on `computeSparseVectorCovariance`.
+    if (!rows.filter(_.isInstanceOf[DenseVector]).isEmpty()) {
```

Comment:
How about `rows.exists(_.isInstanceOf[DenseVector])`? might short-circuit faster. This could be expensive I guess if it really is all sparse, but correctness is more important.

Reply:
I think `rows` is an RDD of Vector. There is no `exists` defined in the RDD class. Please correct me if I am wrong.

Also in the above step, we do expensive operations will computing summary, so not sure this filter is expense comparatively.

Reply:
Ah oops, for some reason thinking this was a Seq.